tiledb/sm/serialization/query.cc

```diff
@@ -667,20 +667,24 @@ Status index_read_state_from_capnp(
     const ArraySchema& schema,
     const capnp::ReadStateIndex::Reader& read_state_reader,
     SparseIndexReaderBase* reader) {
-  auto read_state = reader->read_state();
-
-  read_state->done_adding_result_tiles_ =
-      read_state_reader.getDoneAddingResultTiles();
-
+  bool done_reading = read_state_reader.getDoneAddingResultTiles();
   assert(read_state_reader.hasFragTileIdx());
-  read_state->frag_idx_.clear();
+  std::vector<FragIdx> fragment_indexes;
   for (const auto rcs : read_state_reader.getFragTileIdx()) {
     auto tile_idx = rcs.getTileIdx();
     auto cell_idx = rcs.getCellIdx();
 
-    read_state->frag_idx_.emplace_back(tile_idx, cell_idx);
+    fragment_indexes.emplace_back(tile_idx, cell_idx);
   }
 
+  auto read_state = reader->read_state();
```

Comment:
Let's make read_state() return a ReadState&... That way we don't need the new check below. More, eventually we will also construct a full C.41 reader object in deserialization, the read state will be passed into the constructor, not gotten from the reader and set.